Reorder compilation process so `-miopen-affix-params` becomes the first pass.

whchung opened the pull request:
Reorder the lowering pipeline to `-miopen-affix-params` could become the very first pass.

This PR aims to fulfill https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/337

whchung committed:
Initial codes to get -miopen-affix-params be the first pass.

jerryyin commented on mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp:
```diff
@@ -259,10 +259,10 @@ extern "C" MiirStatus miirLowerTuningParams(MiirHandle mlirHandle) {
-  pm.addPass(mlir::miopen::createAffineTransformPass());
   pm.addPass(
       mlir::miopen::createAffixTuningParametersPass(0, 0, handle->perfConfig));
+  pm.addPass(mlir::miopen::createLowerMIOpenOpsStep1Pass());
+  pm.addPass(mlir::miopen::createAffineTransformPass());
```
Can/should those passes fail? If not, we can remove them from the API. This reduce the overhead for the client to do a quick check.

whchung replied:
We might need to introduce `miopen-lowering-step2` here. That is the pass which would actually fail in case we pass in unsuitable tuning parameters.

jerryyin commented on mlir/tools/mlir-miopen-driver/mlir-miopen-driver.cpp:
```diff
@@ -2472,10 +2472,10 @@ static LogicalResult populateKernelLaunchLogic(
-  pm.addPass(mlir::miopen::createAffineTransformPass());
   pm.addPass(mlir::miopen::createAffixTuningParametersPass(blockSize, gridSize,
                                                            perfConfig));
+  pm.addPass(mlir::miopen::createLowerMIOpenOpsStep1Pass());
+  pm.addPass(mlir::miopen::createAffineTransformPass());
```
We can remove those, this pipeline is only used in unit test validation.

whchung replied:
Will do.

jerryyin commented on mlir/lib/Dialect/MIOpen/Transforms/AffixTuningParameters.cpp:
```diff
@@ -46,84 +46,14 @@ struct AffixTuningParameters : public MIOpenOpsAffixTuningParametersPassBase<Aff
-  //           index_t CThreadCopySrcDstVectorReadWriteDim,      - matrix_c_source_dest_vector_read_write_dim attribute
-  //           index_t CThreadCopyDstDataPerWrite>               - matrix_c_dest_data_per_write attribute
-
   func.walk([&](miopen::GridwiseGemmOp op) { affixTuningParametersImpl(op); });
 
   func.walk(
       [&](miopen::GridwiseGemmV2Op op) { affixTuningParametersImpl(op); });
```
Looks like we can remove those because we'll never encouter a gridwisegemm op in this pass.

whchung replied:
In the next iteration they'll be removed.

jerryyin commented on mlir/lib/Dialect/MIOpen/Transforms/AffixTuningParameters.cpp:
```diff
@@ -46,84 +46,14 @@ struct AffixTuningParameters : public MIOpenOpsAffixTuningParametersPassBase<Aff
 void AffixTuningParameters::runOnFunction() {
   FuncOp func = getFunction();
 
-  // In the original C++ implementation, template arguments used by GridwiseGemmOp are:
```
Thanks for cleaning those up.

jerryyin commented on mlir/include/mlir/Dialect/MIOpen/Tuning/ConvContext.h:
```diff
@@ -205,5 +226,42 @@ template <typename T> static ConvolutionContext populateConvContext(T &op) {
           dilationVal, paddingVal, gemmId, dataType};
 }
 
+#define POPULATE_CONV_CONTEXT(T) \
```
Instead of using macro, why not use template as-is, since you are templatized on the op itself?

whchung replied:
The existing template assumes `T` be `GridwiseGemmOp` or `GridwiseGemmV2Op` which doesn't have `filter` / `input` / `output` fields, so I have to resort to specialized templates.

This will be changed in the next iteration.

jerryyin commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -4843,6 +4847,36 @@ void affixBlockwiseCopyAttributes(
   bop->setAttr("bound", b.getArrayAttr(blockwiseCopyBoundsAttr));
 }
 
+template <typename T, typename U>
+void affixGridwiseGemmAttributes(T &convOp, U &gop, OpBuilder &b) {
```
This looks like a clean way to do it!

At the same time we need more testing to make sure those ops does transfer expected attributes as we expect them to.

jerryyin commented on mlir/include/mlir/Dialect/MIOpen/Tuning/ConvContext.h:
```diff
@@ -108,15 +109,23 @@ struct ConvolutionContext : SQLiteSerializable<ConvolutionContext> {
+  } else if (isa<miopen::Conv2DBwdWeightOp>(*op)) {
     opType = miopen::ConvOpType::Conv2DBwdWeightOpType;
   } else {
-    opType = miopen::ConvOpType::Conv2DOpType;
+    auto kernel_algorithm =
+        op->template getAttrOfType<StringAttr>("kernel_algorithm").getValue();
+    if (kernel_algorithm.find(StringRef("backward_data")) != StringRef::npos) {
+      opType = miopen::ConvOpType::Conv2DBwdDataOpType;
+    } else if (kernel_algorithm.find(StringRef("backward_weight")) !=
+               StringRef::npos) {
+      opType = miopen::ConvOpType::Conv2DBwdWeightOpType;
+    } else {
+      opType = miopen::ConvOpType::Conv2DOpType;
+    }
```
Under what circumstance would this chunk of code be executed? Looks like at this point it isn't either fwd/bwd or wrw.

whchung replied:
I haven't finished changing the unit tests yet so the current implementation allows `miopen-affix-params` be the 1st OR the 3rd pass, hence this logic.

In the next iteration I'll eliminate these logic so `miopen-affix-params` must be the 1st pass in the pipeline.

jerryyin commented on mlir/include/mlir/Dialect/MIOpen/Tuning/ConvContext.h:
```diff
@@ -137,6 +146,18 @@ populateDimVal(const ArrayAttr &layoutAttr, const ArrayAttr &dimAttr,
   }
 }
 
+static inline void
+populateDimVal(const ArrayAttr &layoutAttr, const ArrayRef<int64_t> &dim,
+               llvm::StringMap<std::pair<size_t, int64_t>> &dimIndexVal) {
```
If we are going to use this function on top level conv related op only then I prefer to get rid of the function with the same name above.

whchung replied:
In the next iteration I'll eliminate the function above.

whchung replied:
The function is still needed for C++ emitters. I've added TBD in the comment that we eliminate them along with C++ emitters in the future.

jerryyin replied:
> The function is still needed for C++ emitters.

Good to know, thanks.

whchung committed:
Fix clang-format.

whchung committed:
Reorder -miopen-affix-params to the 1st pass for unit tests.

whchung committed:
Logic to force -miopen-affix-params always be the first pass.

whchung committed:
Change lowering pipeline per review comments.

whchung committed:
Fix clang-format.

whchung committed:
Fix passes lost in earlier commits.

whchung committed:
Fix 3 unit tests which misses -miopen-affix-params pass.

Now -miopen-affix-params must be specified in the beginning of the
lowering pipeline.

whchung committed:
Fix lowering_top_level.mlir .

Test config for miopen.conv2d_bwd_data supplied in this test case
couldn't pass default heuristic tuning parameters used in
-miopen-affix-params, change them to supported ones.

whchung committed:
Fix more miopen.conv2d_bwd_data related unit tests.

whchung committed:
Fix C++ path related unit tests.

jerryyin commented on mlir/tools/mlir-miopen-lib/mlir-miopen-lib.cpp:
```diff
@@ -259,10 +261,11 @@ extern "C" MiirStatus miirLowerTuningParams(MiirHandle mlirHandle) {
   // Passes for lowering MIOpen dialect.
-  pm.addPass(mlir::miopen::createLowerMIOpenOpsStep1Pass());
-  pm.addPass(mlir::miopen::createAffineTransformPass());
   pm.addPass(
       mlir::miopen::createAffixTuningParametersPass(0, 0, handle->perfConfig));
+  pm.addPass(mlir::miopen::createLowerMIOpenOpsStep1Pass());
+  pm.addPass(mlir::miopen::createAffineTransformPass());
+  pm.addPass(mlir::miopen::createLowerMIOpenOpsStep2Pass());
```
Now we can actually need to lower more pipelines to be able to tell if it is going to succeed? From MIOpen perspective they want to keep the applicability check absolutely minimal because that was on the critical pass. Is there any way to remain the same amount of lowering as before.

Also I'd kick off a MIOpen integration round of test on this PR when it is ready just to make sure. Those changes will have impact on it.

jerryyin replied:
BTW, once you've done that, could you add a few test cases (Doesn't have to be in this PR) that make sure certain configs to make either of the above passes to fail. This is just to make sure a `signalPassFailure()` in either of those passes can be propagated correctly to API as a `MIIR_BUILD_FAILURE`

whchung replied:
With upcoming KPACK logic, we need step2 to tell whether we have enough data to compute along GemmK dimension.
And step2 actually doesn't take too much time, compared with LLVM middle-end (opt) or backend (llc).

Additional tests shall likely be added in the KPACK PR.

jerryyin replied:
> And step2 actually doesn't take too much time

Sounds good.

> Additional tests shall likely be added in the KPACK PR.

Is there a way to track it in a ticket? This is pretty important because I want to make sure we reject incorrect configs properly, otherwise bad things can happen from our client.

whchung replied:
It will be covered in https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/199

jerryyin replied:
Documented [here](https://github.com/ROCmSoftwarePlatform/llvm-project-private/issues/199#issuecomment-928257850).

jerryyin commented on mlir/test/mlir-miopen-driver/sanity.mlir:
```diff
@@ -4,53 +4,53 @@
-// RUN: mlir-miopen-driver -p -miopen-lowering -miopen-affine-transform -miopen-affix-params -miopen-lowering-step2 -miopen-lowering-step3 -miopen-lowering-step4 -miopen-lowering-step5 -convert-miopen-to-gpu -convert-gpu-to-rocdl | mlir-translate -mlir-to-rocdlir
-// RUN: mlir-miopen-driver -p -miopen-lowering -miopen-affine-transform -miopen-affix-params -miopen-lowering-step2 -miopen-lowering-step3 -miopen-lowering-step4 -miopen-lowering-step5 -convert-miopen-to-gpu -convert-gpu-to-rocdl | mlir-translate -mlir-to-rocdlir | opt -O3 -S -strip
-// RUN: mlir-miopen-driver -p -miopen-lowering -miopen-affine-transform -miopen-affix-params -miopen-lowering-step2 -miopen-lowering-step3 -miopen-lowering-step4 -miopen-lowering-step5 -convert-miopen-to-gpu -convert-gpu-to-rocdl | mlir-translate -mlir-to-rocdlir | opt -O3 -S -strip | llc -mcpu=gfx900
+// RUN: mlir-miopen-driver -p -miopen-affix-params | mlir-opt
```
Out of curiosity, I remember that when we talked offline, the `-p` wouldn't work with `-miopen-affix-params` pass because it generated a set of incoherent tuning parameters. Since I don't see you making any changes on default parameters on `-p`, how do you make it pass now?

whchung replied:
@jerryyin Forward convolution works. The only problem is on backward data.

jerryyin replied:
Gotcha, from test perspective fwd/bwd/wrw doesn't matter much in first 3 passes.

jerryyin commented on mlir/include/mlir/Dialect/MIOpen/utility/common.h:
```diff
@@ -7,4 +7,7 @@ struct ArgumentFields {
   StringRef gemmTargetCharName[3];
```
Just a random side comment, I actually didn't know we have a `common.h` that get pulled in by both:
 - LowerMIOpenOps.h
 - BackwardWeightV4R4Helper.h

Those seem to only deal with a particular lowering pass and is not that "common". I think when time is right we should work on refactoring to split the `LowerMIOpenOps.h` to per pass based. Then get rid of those helper/common headers.

jerryyin reviewed (commented):
Overall looking very good just some minor comments, thanks for your work!

whchung commented:
Pending on #391 be merged first.

whchung commented:
@jerryyin this PR was passing earlier today, but after rebase to the latest LLVM tip it's failing now.

From CI log it appears to be a node-specific issue though (ixt-rack-11 is being disconnected). Should I just ask jenkins to retest, or should I wait for the node to become online again?

jerryyin commented:
@whchung I have no idea

@okakarpa Could you take a look at what's going on with `ixt-rack-11`

whchung committed:
Fix clang-format.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-reorder-passes
